ParallelPrimitives/RadixSortKernels.h

```diff
@@ -521,278 +514,708 @@ extern "C" __global__ void SortSinglePassKernel( int* gSrcKey, int* gDstKey, int
+		uint32_t x = sMemIO[threadIdx.x];
 
-		int localThreadSum = 0;
-		for( int i = 0; i < N_WGS_EXECUTED; ++i )
+		if( offset <= threadIdx.x )
 		{
-			int current = gCount[binId * N_WGS_EXECUTED + i];
-			gCount[binId * N_WGS_EXECUTED + i] = localThreadSum;
-
-			localThreadSum += current;
+			x += sMemIO[threadIdx.x - offset];
 		}
 
-		// Store the thread local sum to LDS.
+		__syncthreads();
+
+		sMemIO[threadIdx.x] = x;
 
-		blockBuffer[binId] = localThreadSum;
+		__syncthreads();
 	}
+	uint32_t sum = sMemIO[NThreads - 1];
 
-	LDS_BARRIER;
+	__syncthreads();
 
-	// Do parallel exclusive scan on the LDS
+	sMemIO[threadIdx.x] += prefix - value;
 
-	int globalSum = 0;
-	for( int binId = 0; binId < BIN_SIZE; binId += WG_SIZE * 2 )
-	{
-		int* globalOffset = &blockBuffer[binId];
-		int currentGlobalSum = ldsScanExclusive( globalOffset, WG_SIZE * 2 );
-		globalOffset[threadIdx.x * 2] += globalSum;
-		globalOffset[threadIdx.x * 2 + 1] += globalSum;
-		globalSum += currentGlobalSum;
-	}
+	__syncthreads();
 
-	LDS_BARRIER;
+	return sum;
+}
 
-	// Add the global offset to the global histogram.
+extern "C" __global__ void gHistogram( RADIX_SORT_KEY_TYPE* inputs, uint32_t numberOfInputs, uint32_t* gpSumBuffer, uint32_t startBits, uint32_t* counter )
+{
+	__shared__ uint32_t localCounters[sizeof( RADIX_SORT_KEY_TYPE )][256];
 
-	for( int binId = threadIdx.x; binId < BIN_SIZE; binId += WG_SIZE )
+	for( int i = 0; i < sizeof( RADIX_SORT_KEY_TYPE ); i++ )
 	{
-		for( int i = 0; i < N_WGS_EXECUTED; ++i )
+		for( int j = threadIdx.x; j < 256; j += GHISTOGRAM_THREADS_PER_BLOCK )
 		{
-			gHistogram[binId * N_WGS_EXECUTED + i] += blockBuffer[binId];
+			localCounters[i][j] = 0;
 		}
 	}
-}
 
-extern "C" __device__ void WorkgroupSync( int threadId, int blockId, int currentSegmentSum, int* currentGlobalOffset, volatile int* gPartialSum, volatile bool* gIsReady )
-{
-	if( threadId == 0 )
+	__syncthreads();
+
+	uint32_t numberOfBlocks = div_round_up( numberOfInputs, GHISTOGRAM_ITEM_PER_BLOCK );
+	__shared__ uint32_t iBlock;
+	if( threadIdx.x == 0 )
+	{
+		iBlock = atomicInc( counter, 0xFFFFFFFF );
+	}
+
+	__syncthreads();
+
+	bool hasData = false;
+
+	while( iBlock < numberOfBlocks )
 	{
-		int offset = 0;
+		hasData = true;
 
-		if( blockId != 0 )
+#if defined( KEY_IS_16BYTE_ALIGNED )
+		if( ( iBlock + 1 ) * GHISTOGRAM_ITEM_PER_BLOCK <= numberOfInputs )
 		{
-			while( !gIsReady[blockId - 1] )
+			for( int i = 0; i < GHISTOGRAM_ITEM_PER_BLOCK; i += GHISTOGRAM_THREADS_PER_BLOCK * 4 )
 			{
+				uint32_t itemIndex = iBlock * GHISTOGRAM_ITEM_PER_BLOCK + i + threadIdx.x * 4;
+				struct alignas( 16 ) Key4
+				{
+					RADIX_SORT_KEY_TYPE xs[4];
+				};
+				Key4 key4 = *(Key4*)&inputs[itemIndex];
+				for( int k = 0; k < 4; k++ )
+				{
+					auto item = key4.xs[k];
+					for( int i = 0; i < sizeof( RADIX_SORT_KEY_TYPE ); i++ )
+					{
+						uint32_t bitLocation = startBits + i * 8;
+						uint32_t bits = ( getKeyBits( item ) >> bitLocation ) & 0xFF;
```

Comment:
Since this appears many times, please use `getMaskedBits` to make it more readable.

It's from the old code that you've  commented out.

`__device__ constexpr u32 getMaskedBits( const u32 value, const u32 shift ) noexcept { return ( value >> shift ) & RADIX_MASK; }`

Reply:
I added a function extractDigit as getMaskedBits sounds slightly vague to me.